### ISSUE TITLE
tf.io.gfile.glob missing some patterns. Using tf-nightly

### ISSUE THREAD

**Issue by Eshan-Agarwal** [tensorflow/tensorflow#37758]
**System information** 
- Have I written custom code (as opposed to using a stock
example script provided in TensorFlow): 
- OS Platform and Distribution (e.g.,
Linux Ubuntu 16.04):  Windows 10
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if
the issue happens on mobile device: 
- TensorFlow installed from (source or
binary): - TensorFlow version (use command below): tf-nightly
- Python version: - Bazel
version (if compiling from source):
- GCC/Compiler version (if compiling from
source): 
- CUDA/cuDNN version: - GPU model and memory:

**Describe the current behavior**
cc @Conchylicultor,
Please have a look on issue from `TFDS` tensorflow/datasets#1670, tests are failing for `PlantVillage `and `The300wLp` datasets because in `_generate_example ` function of both [`plant_village.py`](https://github.com/tensorflow/datasets/blob/master/tensorflow_datasets/image/plant_village.py#L142) and [the300w_lp.py](https://github.com/tensorflow/datasets/blob/master/tensorflow_datasets/image/the300w_lp.py#L116) `tf.io.gfile.glob()` does not correctly matches all examples patterns. However python [`glob`](https://docs.python.org/3/library/glob.html) solves issue see PR tensorflow/datasets#1684
**Describe the expected behavior**
tf.io.gfile.glob() must matches all patterns provided so that all required examples are generated.

**Standalone code to reproduce the issue** 
Please have a look on this [`colab`](https://colab.research.google.com/drive/1tfLOQubRWd6Dc9mPTtAwca-eAzctGLId) notebook, it contains all tracebacks as well as problem with ` tf.io.gfile.glob()` and how python `glob `solves this issue.

**As glob fix this issue but we have to use `tf.io.gfile` because we need to support GCS and other distributed files systems.**

**Comment by Conchylicultor** [comment]
Yes, TFDS tests have started failing for patterns like: `tf.io.gfile.glob('/path/to/file/[!Code]*[!_Flip]/[!_]*.jpg')` or `tf.io.gfile.glob('/path/to/*.[jJ][pP][gG]')`.

@Eshan-Agarwal Could you provide a small self-contained code snippet to reproduce the issue ?

Something like:
```
import glob
import tensorflow as tf

with tf.io.gfile.GFile('/tmp/file') as f:
  pass

print(list(tf.io.gfile('/tmp/some_pattern')))
print(list(glob.glob('/tmp/some_pattern')))   # Should show different result
```

**Comment by mihaimaruseac** [comment]
+1 on providing simple patterns.

I can look into this in about a month or two, once modular filesystems (tensorflow/community#101) are implemented

**Comment by Eshan-Agarwal** [comment]
@Conchylicultor  @mihaimaruseac  please look on this [colab](https://colab.research.google.com/drive/1rXrpPVhiDfUH4DBT3Og65yx61R5MleHr) notebook

**Comment by Conchylicultor** [comment]
@Eshan-Agarwal I get `NotFoundError: /content/temp/plant_village/Grape___Leaf_blight_(Isariopsis_Leaf_Spot); No such file or directory` when executing your colab.

**Comment by mihaimaruseac** [comment]
@Eshan-Agarwal the difference between the colab and the example template suggested is that we need to have the exact same setup for the colab, whereas the suggested template creates the files (with zero bytes) so it can be easily converted into a test case that now fails and after fixing will succeed.

But it's ok, I'll take care of this issue.

**Comment by Conchylicultor** [comment]
@Eshan-Agarwal For the future, here is what a minimum reproductible example looks like:

```python
import os
import glob
import tensorflow.compat.v2 as tf

# Write a dummy file
root_dir = '/tmp/dir_with_(brace)/'
tf.io.gfile.makedirs(root_dir)
with tf.io.gfile.GFile(os.path.join(root_dir, 'some_file.txt'), 'w') as f:
  f.write('')

# Search the file
glob_path = os.path.join(root_dir, "*")
print(list(glob.iglob(glob_path)))        # ['/tmp/dir_with_(brace)/some_file.txt']
print(list(tf.io.gfile.glob(glob_path)))  # []
```

This allow the team to easily undersand what the issue is.

@mihaimaruseac here, it seems that `tf.io.gfile.glob` fails when `(` are present in the path. This is a regression as it only appear in TF nightly. Not TF 2.1

**Comment by mihaimaruseac** [comment]
I think I have an idea where that might come from.

**Comment by Eshan-Agarwal** [comment]
@Conchylicultor  @mihaimaruseac  thanks for your quick responses, Actually I upload temp folder containing some example you can download folder from [here](https://drive.google.com/drive/folders/1DyMHIXSUue10k1a9pDbwgHK0Lng9VxuM?usp=sharing).
but it is good to use code provided by @Conchylicultor  without any external uploading.

**Comment by Conchylicultor** [comment]
@mihaimaruseac, do you know if this will this be fixed before the 2.2 release ? Otherwise, it will break some of our TFDS users using those datasets.
Worse, as no error is raised with `tf.io.gfile.glob`, it will silently generate bad datasets where some examples are missing.

@gadagashwini I'm not sure why you added the `TF 2.1` label. `TF 2.1` works fine, but `tf-nightly` is broken.

**Comment by mihaimaruseac** [comment]
We haven't prioritized this for 2.2 release, unfortunately. Can you verify that installing 2.2.-rc1 breaks you? Let me know if that is the case and I'll see if we can prioritize it.

**Comment by Conchylicultor** [comment]
@Eshan-Agarwal, could you check if the issue occurs with `2.2.-rc1` ?

**Comment by Eshan-Agarwal** [comment]
@Conchylicultor  sure, I will

**Comment by Eshan-Agarwal** [comment]
@Conchylicultor  @mihaimaruseac Yes same problem occurs with `2.2.-rc1`, `tf.io.gfile.glob()` not matches patterns. 
Please see this [colab ](https://colab.research.google.com/drive/1SAXEnHx0fXdnRjKYRdk3g0BzCoCgitKL)notebook, it shows both original tests as well as test with minimal reproductible code.

**Comment by Conchylicultor** [comment]
@Eshan-Agarwal thank you for confirming.

@mihaimaruseac I believe this should be prioritised. This not only impact TFDS but potentially every users using `tf.io.gfile.glob`. As the issue is silent, users may not even notice there is a bug.
In our case we got lucky to have good unit-tests.
Note: The issue only happened externally. Internally, our tests works fine.

**Comment by mihaimaruseac** [comment]
Reopening to close once the fix lands on `r2.2` branch (and then will be picked up by the next RC).

See #37915

**Comment by Conchylicultor** [comment]
@mihaimaruseac Thank you for the fix.
I'll update here when [our tests](https://storage.googleapis.com/tfds-kokoro-public/kokoro-build.html) are back to green.

**Comment by goldiegadde** [comment]
the cherrypick has been merged into the r2.2 branch

**Comment by mihaimaruseac** [comment]
TF 2.2.0-rc2 has been released and this issue should be fixed now.

**Comment by Conchylicultor** [comment]
I confirm this fixed our tests. Thank you very much!
